programs/kyupad-ido/src/instructions/invest.rs

```diff
@@ -0,0 +1,199 @@
+use anchor_lang::prelude::*;
+use spl_token::{
+    instruction::transfer_checked,
+    solana_program::{self, program::invoke, system_instruction},
+};
+
+use crate::*;
+
+use spl_associated_token_account::get_associated_token_address_with_program_id;
+
+use anchor_spl::token::Mint;
+
+use self::{
+    errors::KyuPadError,
+    utils::{assert_keys_equal, verify},
+};
+
+pub fn invest<'c: 'info, 'info>(
+    ctx: Context<'_, '_, 'c, 'info, Invest<'info>>,
+    invest_args: InvestArgs,
+) -> Result<()> {
+    let project_counter: &mut Account<ProjectCounter> = &mut ctx.accounts.project_counter;
+    let project = &ctx.accounts.project;
+    let investor_counter = &mut ctx.accounts.investor_counter;
+    let investor = &ctx.accounts.investor;
+    let investment_destination = &ctx.accounts.investment_destination;
+    let system_program = &ctx.accounts.system_program;
+
+    // check if has a valid max_ticket_amount
+    if invest_args.ticket_amount > invest_args.max_ticket_amount {
+        return Err(KyuPadError::InvalidTotalInvestment.into());
+    }
+
+    // Check if in the time allow
+    let clock = Clock::get()?;
+    let current_timestamp = clock.unix_timestamp;
+    if current_timestamp < project.start_date || current_timestamp > project.end_date {
+        return Err(KyuPadError::NotInvestTime.into());
+    }
+
+    // check if have enough ticket to invest
+    if project_counter.remainning <= 0 {
+        return Err(KyuPadError::ProjectOutOfTicket.into());
+    }
+
+    let binding = investor_counter.to_account_info();
+    let mut __data: &[u8] = &binding.try_borrow_data()?;
+    let mut __disc_bytes = [0u8; 8];
+    __disc_bytes.copy_from_slice(&__data[..8]);
+    let __discriminator = u64::from_le_bytes(__disc_bytes);
+
+    // check if user is enough ticket to invest
+    if __discriminator != 0 {
+        if invest_args.ticket_amount > investor_counter.remainning {
+            return Err(KyuPadError::NotEnoughTicket.into());
+        }
+    } else {
+        investor_counter.remainning = invest_args.max_ticket_amount;
+        investor_counter.wallet = *investor.key;
+        investor_counter.project_id = invest_args.project_id;
```

Comment:
Why we saving this?

Reply:
To mark pda investor_counters to help fetch those who have invested in this project instead of having to save an array of people who have invested